src/Microsoft.Management.Deployment/PackageCatalogProgress.cpp

```diff
@@ -0,0 +1,137 @@
+            {
+                std::unordered_map<AppInstaller::ProgressType, double> progressWeights;
+
+                // There is no download operation for remove operation, so remove the download progress.
+                if (removeOperation)
+                {
+                    // it is percentage based progress.
+                    progressWeights.insert_or_assign(AppInstaller::ProgressType::Bytes, 0);
+                }
+                else
+                {
+                    // Add/Update operation has two progress types:
+                    // 1. Bytes for downloading index and
+                    // 2. Percent for index installation.
+                    progressWeights.insert_or_assign(AppInstaller::ProgressType::Bytes, 0.7);
+                    progressWeights.insert_or_assign(AppInstaller::ProgressType::Percent, 0.3);
+                }
+
+                return std::make_shared<PreIndexedPackageCatalogProgressSink>(progressWeights, progressReporter);
+            }
+            else
+            {
+                return std::make_shared<CompletionOnlyProgressSink>(progressReporter);
+            }
+        }
+    }
+
+    CompletionOnlyProgressSink::CompletionOnlyProgressSink(std::function<void(double)> progressReporter) :
+        m_progressReporter(progressReporter)
+    {
+    }
+
+    void CompletionOnlyProgressSink::OnProgress(uint64_t current, uint64_t maximum, AppInstaller::ProgressType type)
+    {
+        UNREFERENCED_PARAMETER(current);
+        UNREFERENCED_PARAMETER(maximum);
+        UNREFERENCED_PARAMETER(type);
+    }
+
+    void CompletionOnlyProgressSink::SetProgressMessage(std::string_view message)
+    {
+        UNREFERENCED_PARAMETER(message);
+    }
+
+    void CompletionOnlyProgressSink::BeginProgress()
+    {
+        m_progressReporter(0);
+    }
+
+    void CompletionOnlyProgressSink::EndProgress(bool hideProgressWhenDone)
+    {
+        UNREFERENCED_PARAMETER(hideProgressWhenDone);
+        m_progressReporter(100);
+    }
+
+    PreIndexedPackageCatalogProgressSink::PreIndexedPackageCatalogProgressSink(std::unordered_map<AppInstaller::ProgressType, double> progressWeights, std::function<void(double)> progressReporter) :
+        m_progressWeights(progressWeights), m_progressReporter(progressReporter)
+    {
+        // If no weights are provided, default to percent.
+        if (m_progressWeights.empty())
+        {
+            m_progressWeights[AppInstaller::ProgressType::Percent] = 1;
+        }
+
+        // Calculate the total weight.
+        double totalWeight = 0;
+        for (const auto& weight : m_progressWeights)
+        {
+            if (weight.first != AppInstaller::ProgressType::None)
+            {
+                totalWeight += weight.second;
+            }
+        }
+
+        // If the total weight is greater than 1, throw an exception.
+        if (totalWeight != 1.0)

```

Comment:
I might be missing something, but it seems like the remove operation code will hit this check based on only setting bytes to 0 in the factory.

Reply:
I made a mistake in the logic. My intention was to include only the percentage progress type with a weight of 1.0 for the remove operation, but I incorrectly set it to 0. It is addressed now.